posthog/caching/fetch_from_cache.py

```diff
@@ -48,23 +57,70 @@ def fetch_cached_insight_result(target: Union[Insight, DashboardTile]) -> Insigh
         return NothingInCacheResult(cache_key=cache_key)
     else:
         statsd.incr("posthog_cloud_insight_cache_hit")
+        last_refresh = cached_result.get("last_refresh")
+        next_allowed_refresh = cached_result.get("next_allowed_refresh") or last_refresh + refresh_frequency
+
         return InsightResult(
             result=cached_result.get("result"),
-            last_refresh=cached_result.get("last_refresh"),
+            last_refresh=last_refresh,
             cache_key=cache_key,
             is_cached=True,
             # :TODO: This is only populated in some code paths writing to cache
             timezone=cached_result.get("timezone"),
+            next_allowed_refresh=next_allowed_refresh,
         )
 
 
-def synchronously_update_cache(insight: Insight, dashboard: Optional[Dashboard]) -> InsightResult:
+def synchronously_update_cache(
+    insight: Insight, dashboard: Optional[Dashboard], refresh_frequency: Optional[timedelta] = None
+) -> InsightResult:
     cache_key, cache_type, result = calculate_result_by_insight(team=insight.team, insight=insight, dashboard=dashboard)
     timestamp = now()
+
+    next_allowed_refresh = timestamp + refresh_frequency if refresh_frequency else None
     update_cached_state(
-        insight.team_id, cache_key, timestamp, {"result": result, "type": cache_type, "last_refresh": timestamp}
+        insight.team_id,
+        cache_key,
+        timestamp,
+        {"result": result, "type": cache_type, "last_refresh": timestamp, "next_allowed_refresh": next_allowed_refresh},
     )
 
     return InsightResult(
-        result=result, last_refresh=timestamp, cache_key=cache_key, is_cached=False, timezone=insight.team.timezone
+        result=result,
+        last_refresh=timestamp,
+        cache_key=cache_key,
+        is_cached=False,
+        timezone=insight.team.timezone,
+        next_allowed_refresh=next_allowed_refresh,
+    )
+
+
+# returns should_refresh, refresh_frequency
+def should_refresh_insight(insight: Insight, dashboard_tile: Optional[DashboardTile]) -> Tuple[bool, timedelta]:
```

Comment:
The signature here is a bit weird.

There are two patterns already in `posthog/caching`:
- `def fetch_cached_insight_result(target: Union[Insight, DashboardTile]) `
- `def synchronously_update_cache(insight: Insight, dashboard: Optional[Dashboard])`

Could we have a signature that's either of these rather than a new thing? Makes callsites more uniform.

Reply:
I thought about this a bit as well but the existing signatures don't mix too well with what needs to be done here. The `target` approach makes for some annoying handling regarding getting the dashboard filters and the other approach makes it annoying to get the tile in context. 

But yeah, I'll think of a way to standardize in a way that's not too unclean